### ISSUE TITLE
metadata-inspired HttpConnectionManager refactors

### ISSUE THREAD

**Issue by alyssawilk** [envoyproxy/envoy#5842]
tl;dr #5656 inspired a few of us I to take a look at the HCM to figure out if we could add generic useful features to the HCM to work around the issues described in the review of that PR, or if we should just do one-off complexity in the metadata handling.

We're leaning towards extending the filter return codes to allow for more flexibility, as well as debating how to handle filter "callbacks" which most filter implementer don't care about (e.g. metadata, 100 continue).    Options that we've come up with thusfar are documented here:

https://docs.google.com/document/d/16istepO8qfG09UYdgxVzCOwJcjVL9Ua_AtcwJPyJUGI/edit?ts=5c59a7cf

(comments should be open to the public unless I screwed up permissions).  If you're interested in the internals of how the HTTPConnectionManager manages streams and communicates to filters, please weigh, in as we'd like to sort out a plan and start hacking away :-)

**Comment by alyssawilk** [comment]
@lizan @mattklein123 @soya3129

**Comment by alyssawilk** [comment]
Ok, so I think the current plan is to go with the StopAllIteration filter code for at least headers (unclear if we need it for bodies as well) and losing the pure virtual for filter callbacks.

@soya3129 are you up for tackling these?  I think if we do this, then refactor the relevant filters, and disable metadata by default, we can come back to your latest metadata PR without adding StopIteration to it.

**Comment by mattklein123** [comment]
> I am very excited to give it a try! Thanks!

Thank you @soya3129 for being patient while we sort this out and being willing to work through this!

**Comment by lizan** [comment]
> I think if we do this, then refactor the relevant filters, and disable metadata by default

Thank you, the plan sounds good to me.

**Comment by alyssawilk** [comment]
+1, thanks for picking this up, @soya3129 

Bonus points if you come up with a better name than StopAllIteration if for no other reason than I keep reading this as Stop Alliteration ;-)
